tools/codegen/selective_build/operator.py

```diff
@@ -0,0 +1,117 @@
+    # 'include_all_overloads' flag in this class.
+    name: str
+
+    # True if this is a root operator (i.e. called directly from TorchScript, etc...)
+    is_root_operator: bool
+
+    # Is this operator used for on-device training? If True, then we need to
+    # use the information to generate code in VariableType_N.cpp for registration
+    # of training related operators
+    is_used_for_training: bool
+
+    # If True, it indicates that this operator instance (object) refers to an
+    # operator without the overload name and should apply to all overloads
+    # which have this operator name as the base name. This flag is applicable
+    # only for objects that have operator names without a DOT (period) character
+    # in them.
+    include_all_overloads: bool
+
+    # The list of models that use this operator.
+    models: Optional[List[PyTorchModelMetadata]]
+
+    @staticmethod
+    def from_yaml_dict(op_name: str, op_info: Dict[str, object]) -> 'SelectiveBuildOperator':
+        is_root_operator = op_info.get('is_root_operator', True)
+        assert isinstance(is_root_operator, bool)
+
+        is_used_for_training = op_info.get('is_used_for_training', True)
+        assert isinstance(is_used_for_training, bool)
+
+        include_all_overloads = op_info.get('include_all_overloads', True)
+        assert isinstance(include_all_overloads, bool)
+
+        models: Optional[List[PyTorchModelMetadata]] = None
+        if 'models' in op_info:
+            models_list = op_info['models']
+            assert isinstance(models_list, list)
+            models = list(map(
+                lambda x: PyTorchModelMetadata.from_yaml(x),
+                models_list,
+            ))
+
+        return SelectiveBuildOperator(
+            op_name,
+            is_root_operator,
+            is_used_for_training,
+            include_all_overloads,
+            models,
+        )
+
+    @staticmethod
+    def from_legacy_operator_name_without_overload(name: str) -> 'SelectiveBuildOperator':
+        return SelectiveBuildOperator(
+            name,
+            True,
+            True,
+            True,
+            None,
+        )
+
+def combine_operators(
+        lhs: 'SelectiveBuildOperator',
+        rhs: 'SelectiveBuildOperator') -> 'SelectiveBuildOperator':
+    if str(lhs.name) != str(rhs.name):
+        raise Exception(
+            "Expected both arguments to have the same name, but got '{}' and '{}' instead".format(
+                str(lhs.name),
+                str(rhs.name),
+            )
+        )
+
+    models = None
+    if lhs.models is not None and rhs.models is not None:
+        models = lhs.models[:] + rhs.models[:]
+    elif lhs.models is not None:
+        models = lhs.models[:]
+    elif rhs.models is not None:
+        models = rhs.models[:]
+
+    return SelectiveBuildOperator(
+        lhs.name,
+        lhs.is_root_operator or rhs.is_root_operator,
+        lhs.is_used_for_training or rhs.is_used_for_training,
```

Comment:
At first glance, merging these is_xxx properties seems a bit confusing - why a op used in both non-root context and root context should keep the is_root_operator flag? It would be easier if it is called 'include_root_op_wrapper' so that we know these things are additive (instead of restrictive), i.e. because it's root op we are generating more code, rather than because it's root op we are generating less code.

Another way to do merging is to allow multiple 'SelectiveBuildOperator' instances associated with one operator name. e.g.:
1. "aten::xyz.overload1 -> include_root_wrapper"
2. "aten::xyz.*"
3. "aten::xyz.overload2 -> include_autograd"
4. "aten::xyz.overload2 -> include_root_wrapper"
It's kinda hard to merge these all into one entry "aten::xyz.* -> include_root_wrapper & include_autograd", which will be a superset instead of exact match.

It's a bit like disjunctive normal form of bool expression, each operator filter instance is a minterm that doesn't overlap with other operator filters. Maybe the only exception is 'aten::xyz.*' (include all ops case) which can cause problem when merging with other filters, e.g.: merging 3 and 4 in the example above is probably fine, but merging 2, 3 and 4 is changing the logic. We can keep 2 as a separate term and deprecate it after migrating to the world where overload name cannot be omitted.

Reply:
We can keep calling it "is_root_operator" / "is_used_for_training" as well, but should note it somewhere the way we merge them kinda depends on their downstream usage.

Reply:
@ljk53 Thanks! I've commented this in the most recent revision both the the class definition as well as in the merging code to make it clear what the semantics of it are - i.e. that it applies at the app. level and that an op. is considered to be a root operator if it appears as a root op. in any of the models used in this app.

Reply:
@ljk53 I understand your example. i.e. suppose we annotate `aten::xyz` as a root op, then it marks all the overloads as a root op - the way it's currently implemented. However, there is another control knob `include_all_overload`, so only if that flag is set, will it apply to all overloads. In addition, this will be a problem only for models that use the current static selective build and shouldn't cause "over-fetching" for tracing based operator selection.